### PR TITLE
payment: perform AUTO_PAY_OFF check early when paying invoices

### PR DESCRIPTION
**Pull request by pierre**
This fixes https://github.com/killbill/killbill/issues/812, https://github.com/killbill/killbill/issues/1193 and https://github.com/killbill/killbill/issues/1211.

### CODE REVIEW

**Comment by pierre** [comment]
@sbrossie Assuming CircleCi is green, this should be ready to be reviewed now.

File: jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java [hunk 1]
```diff
@@ -352,14 +347,27 @@ public Response createSubscriptionsWithAddOnsInternal(final List<BulkSubscriptio
...
+
+            private boolean isImmediateOp = true;
+
             @Override
             public List<UUID> doOperation(final CallContext ctx) throws EntitlementApiException {
+                for (final BaseEntitlementWithAddOnsSpecifier spec: baseEntitlementWithAddOnsSpecifierList) {
+                    if (spec.getBillingEffectiveDate() != null) {
+                        final LocalDate nowInAccountTimeZone = new LocalDate(callContext.getCreatedDate(), spec.getBillingEffectiveDate().getChronology().getZone());
+                        if (spec.getBillingEffectiveDate().isAfter(nowInAccountTimeZone)) {
+                            isImmediateOp = false;
```
Comment (sbrossie) [inline jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java]: I am questioning whether this logic will lead to the right expectation:
* `isImmediateOp` = `false` => this was obtained by checking that `spec#getBillingEffectiveDate` is not null and is  in the future, all good.
* `isImmediateOp` = `true` => obtained as a default for when prev condition is not met. However, prev condition only checked that `spec#getBillingEffectiveDate` is not null and is  in the future, but we could still have `spec.getBillingEffectiveDate().is**Not**After(nowInAccountTimeZone)` and yet have a date that is in the future ?

File: jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java [hunk 2]
```diff
@@ -351,15 +352,38 @@ public Response createSubscriptionsWithAddOnsInternal(final List<BulkSubscriptio
             baseEntitlementWithAddOnsSpecifierList.add(baseEntitlementSpecifierWithAddOns);
         }
 
+
+        // Retrieve the tags for the ACCOUNT object to correctly implement callCompletion in the simple use-cases.
+        // In reality, this is much more complex though and not all scenarii are supported (e.g. entitlement plugin could add some tags on the fly).
+        final List<Tag> accountTags = tagUserApi.getTagsForAccountType(account.getId(), ObjectType.ACCOUNT, false, callContext);
```
Comment (sbrossie) [inline jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java]: Should we only fetch those when `callCompletion` = `true` ?

File: jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java (hunk 1 above)
Comment (pierre) [reply jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java]: Good catch! See 60f3dfe3028ba99abc4859c33380042102e1e7cb.

The only downside is that I had to re-use some of the logic internal to `InternalCallContextFactory` inside jaxrs.

**Pull request merged**